mlx/ops.cpp

```diff
@@ -2758,4 +2760,91 @@ array dequantize(
   return w_full;
 }
 
+std::map<char, int> string_to_char_map(std::string inp) {
+  std::map<char, int> counts;
+  for (int i = 0; i < inp.size(); i++) {
+    counts[inp[i]] = i;
+  }
+  return counts;
+}
+
+array einsum(
+    std::string equation,
+    const std::vector<array>& operands,
+    StreamOrDevice s /* = {} */) {
+  std::string output;
+  std::string input = equation;
+  if (equation.find("->") == std::string::npos) {
+    output = std::string(equation);
+    std::sort(output.begin(), output.end());
+  } else {
+    auto pos = equation.find("->");
+    output = equation.substr(pos + 2);
+    input = equation.substr(0, pos);
+  }
+  std::vector<std::string> inputs;
+  std::stringstream ss(input);
+  std::string token;
+  while (getline(ss, token, ',')) {
+    inputs.push_back(token);
+  }
+  if (operands.size() != inputs.size()) {
+    throw std::runtime_error(
+        "Number of operands must match the number of input characters");
+  }
+
+  std::map<char, int> input_map;
+  for (int i = 0; i < inputs.size(); i++) {
+    auto arr = operands[i];
+    auto inp = inputs[i];
+    for (int j = 0; j < MIN(arr.shape().size(), inp.size()); j++) {
+      input_map[inp[j]] = arr.shape().at(j);
+    }
+  }
+  std::vector<int> broad;
+  for (auto key : input_map) {
+    broad.push_back(key.second);
+  }
+  std::vector<array> inputs_arr;
+  for (int i = 0; i < operands.size(); i++) {
+    auto arr = operands[i];
+    auto ord_map = string_to_char_map(inputs[i]);
+    std::vector<int> new_shape;
+    for (auto key : input_map) {
+      if (ord_map.find(key.first) != ord_map.end()) {
+        new_shape.push_back(key.second);
+      } else {
+        new_shape.push_back(1);
+      }
+    }
+    std::vector<int> axis;
+    for (auto key : ord_map) {
+      axis.push_back(key.second);
+    }
+    inputs_arr.push_back(
+        broadcast_to(reshape(transpose(arr, axis, s), new_shape, s), broad, s));
+  }
+
+  auto ord_output = string_to_char_map(output);
+  std::vector<int> rhs_order;
+  for (auto key : ord_output) {
+    rhs_order.push_back(key.second);
+  }
+
+  std::vector<int> sum_axis;
+  int i = 0;
+  for (auto key : input_map) {
+    if (ord_output.find(key.first) == ord_output.end()) {
+      sum_axis.push_back(i);
+    }
+    i += 1;
+  }
+
+  auto acc = ones_like(inputs_arr.at(0), s);
```

Comment:
Start with `inputs_arr[0]` rather than include a new array for accumulation. There should be at least one input right?

Reply:
Actually would be good to check for that at the top and throw if it's right, and maybe add a test case that check that we throw for that.

Reply:
strangely this does not work, if i switch it to use the first array and then accumulate it gets some strange results on the test cases

Reply:
Did you ever get to the bottom of this? Let me know I can take a look as well?

Reply:
That would be helpful, I messed with it a little but didn't find anything concrete. 
I tried calling eval on `inputs_arr.at(0)` to see if maybe certain ops didnt take affect but it didn't resolve it